python/ray/_raylet.pyx

```diff
@@ -2732,6 +2737,48 @@ cdef void terminate_asyncio_thread() nogil:
         core_worker = ray._private.worker.global_worker.core_worker
         core_worker.stop_and_join_asyncio_threads_if_exist()
 
+
+cdef void call_actor_shutdown() noexcept nogil:
+    """C++ wrapper function that calls the Python actor shutdown callback."""
+    with gil:
+        _call_actor_shutdown()
+
+
+def _call_actor_shutdown():
+    """Internal function that calls actor's __ray_shutdown__ method."""
+    try:
+        worker = ray._private.worker.global_worker
+
+        if not worker.actors:
+            return
+
+        actor_id, actor_instance = next(iter(worker.actors.items()))
+        if actor_instance is not None:
+            try:
+                actor_instance.__ray_shutdown__()
+            except Exception:
+                logger.exception("Error during actor __ray_shutdown__ method")
+            finally:
+                worker.actors.pop(actor_id, None)
+    except Exception:
+        # Catch any system-level exceptions to prevent propagation to C++
+        logger.exception("System error during actor shutdown callback")
+
+
+def _register_actor_shutdown_callback():
+    """Register callback to shutdown actor instance before shutdown."""
+    worker = ray._private.worker.global_worker
+    core_worker = worker.core_worker
+    actor_id = core_worker.get_actor_id()
+
+    # Only register callback if actor has __ray_shutdown__ method
+    # This preserves backward compatibility: actors without __ray_shutdown__
+    # will use Python's normal exit flow (including atexit handlers)
+    actor_instance = worker.actors.get(actor_id)
+    if actor_instance is not None and hasattr(actor_instance, '__ray_shutdown__') and callable(getattr(actor_instance, '__ray_shutdown__')):
+        core_worker.set_actor_shutdown_callback(_call_actor_shutdown)
```

Comment:
we are implicitly tying the semantics of `_register_actor_shutdown_callback` to the implementation of `_call_actor_shutdown` (the latter assumes that the former only registers if the actor has the method). it would be preferable to always register the shutdown callback and make it handle edge cases appropriately instead (it can check for the dunder method and no-op if it isn't defined).

by the way, can you help me understand the backwards compatibility comment? won't we will go through Python's normal flow and atexit even if the shutdown callback is called?